src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -84,15 +86,50 @@ class UserService(val directoryDAO: DirectoryDAO, val googleDirectoryDAO: Google
     } yield deleteResult
   }
 
+  def createUserPetServiceAccount(user: WorkbenchUser): Future[WorkbenchUserPetServiceAccountEmail] = {
+    def create() = {
+      val (petSa, petSaDisplayName) = toPetSAFromUser(user)
+      for {
+        petServiceAccount <- googleIamDAO.createServiceAccount(petServiceAccountConfig.googleProject, petSa, petSaDisplayName)
+        _ <- googleDirectoryDAO.addMemberToGroup(WorkbenchGroupEmail(toProxyFromUser(user.id.value)), petServiceAccount.email)
+        _ <- Future.traverse(petServiceAccountConfig.serviceAccountActors) { email =>
+          googleIamDAO.addServiceAccountActorRoleForUser(petServiceAccountConfig.googleProject, petServiceAccount.email, email)
+        }
+        _ <- directoryDAO.addPetServiceAccountToUser(user.id, petServiceAccount.email)
+      } yield petServiceAccount.email
```

Comment:
tricky thing: this is dealing with 2 non-transactional systems, ldap and google. What happens when any one of these steps fails? Looks to me like there is a possibility of an orphaned pet. Perhaps some best attempt recovery code to remove the orphan?

Reply:
Good point. Will look at adding some recovery/cleanup code.

Reply:
> orphaned pet

😿

Reply:
I added some recovery code to delete the pet SA in google and disassociate it from the user in LDAP.